code/__HELPERS/abstract_types.dm

```diff
@@ -10,5 +10,12 @@
 	return abstracts
 
 /// Like subtypesof, but automatically excludes abstract typepaths
-/proc/valid_subtypesof(datum/sometype)
-	return subtypesof(sometype) - get_abstract_types()
+/proc/valid_subtypesof(datum/base_type)
+	return valid_typesof(base_type) - base_type
+
+/// Like typesof, but automatically excludes abstract typepaths
+/proc/valid_typesof(datum/base_type)
+	. = list()
+	for (var/datum/some_type as anything in typesof(base_type))
+		if (some_type != some_type::abstract_type)
+			. += some_type
```

Comment:
Why are you writing your own code here when `get_abstract_types()` already did this for you

Ie why not
```dm
/proc/valid_typesof(datum/base_type)
	return typesof(base_type) - get_abstract_types()
```

Reply:
Because gathering a list of every single abstract type in the game and then substracting them from a list of every type of something is a really stupid way of doing this?

Like, yes, it would be less code or whatever, but I think it's a massive code smell when you can just filter through the types with a ridiculously short for loop.

Reply:
(When we were implementing this proc, we did it in this way so you didn't accidentally iterate over 10000s of types twice rather that once)

Reply:
> How are you sure that it's a smell and not someone's optimization?

I went to the original PR this was implemented in and investigated why it was done this way. The reason was simply because that's how it was done the first time around. I.e. I looked at #93541 prior to making the change.

SyncIt made a review on why it was done this way instead of just filtering via a for loop. The reason was... nothing, and the review was ignored, with the author giving a seemingly nonsensical response to the review.

I don't quite understand how you're going to end up iterating over 10000s of types twice rather than once here. It's just one for loop and a check. Technically there are two for loops if you count `typesof()` as a for loop.